diff -up openssh-7.4p1/sshd.c.newline-banner openssh-7.4p1/sshd.c
--- openssh-7.4p1/sshd.c.newline-banner	2017-02-17 14:00:47.237168594 +0100
+++ openssh-7.4p1/sshd.c	2017-02-17 14:02:10.933096707 +0100
@@ -369,15 +369,15 @@ sshd_exchange_identification(struct ssh
 {
 	u_int i;
 	int remote_major, remote_minor;
-	char *s, *newline = "\n";
+	char *s;
 	char buf[256];			/* Must not be larger than remote_version. */
 	char remote_version[256];	/* Must be at least as big as buf. */
 
-	xasprintf(&server_version_string, "SSH-%d.%d-%.100s%s%s%s",
+	xasprintf(&server_version_string, "SSH-%d.%d-%.100s%s%s\r\n",
 	    PROTOCOL_MAJOR_2, PROTOCOL_MINOR_2,
 	    (options.show_patchlevel == 1) ? SSH_VENDOR_PATCHLEVEL : SSH_VERSION,
 	    *options.version_addendum == '\0' ? "" : " ",
-	    options.version_addendum, newline);
+	    options.version_addendum);
 
 	/* Send our protocol version identification. */
 	if (atomicio(vwrite, sock_out, server_version_string,
