diff -up openssh-7.4p1/servconf.c.memory-problems openssh-7.4p1/servconf.c
--- openssh-7.4p1/servconf.c.memory-problems	2017-02-09 10:41:42.483123417 +0100
+++ openssh-7.4p1/servconf.c	2017-02-09 10:42:16.392102462 +0100
@@ -2006,6 +2006,8 @@ copy_set_server_options(ServerOptions *d
 		dst->n = src->n; \
 } while (0)
 
+	u_int i;
+
 	M_CP_INTOPT(password_authentication);
 	M_CP_INTOPT(gss_authentication);
 	M_CP_INTOPT(pubkey_authentication);
@@ -2058,8 +2060,10 @@ copy_set_server_options(ServerOptions *d
 } while(0)
 #define M_CP_STRARRAYOPT(n, num_n) do {\
 	if (src->num_n != 0) { \
+		for (i = 0; i < dst->num_n; i++) \
+			free(dst->n[i]); \
 		for (dst->num_n = 0; dst->num_n < src->num_n; dst->num_n++) \
-			dst->n[dst->num_n] = xstrdup(src->n[dst->num_n]); \
+			dst->n[dst->num_n] = src->n[dst->num_n]; \
 	} \
 } while(0)
 
diff -up openssh-7.4p1/sshd.c.memory-problems openssh-7.4p1/sshd.c
