diff --git a/serverloop.c b/serverloop.c
index b5eb3440..1535eeb2 100644
--- a/serverloop.c
+++ b/serverloop.c
@@ -225,9 +225,10 @@ wait_until_can_do_something(int connection_in, int connection_out,
 		uint64_t keepalive_ms =
 		    (uint64_t)options.client_alive_interval * 1000;
 
-		client_alive_scheduled = 1;
-		if (max_time_ms == 0 || max_time_ms > keepalive_ms)
+		if (max_time_ms == 0 || max_time_ms > keepalive_ms) {
 			max_time_ms = keepalive_ms;
+			client_alive_scheduled = 1;
+		}
 	}
 
 #if 0

