diff --git a/myproposal.h b/myproposal.h
index 1bf072d..6a92111 100644
--- a/myproposal.h
+++ b/myproposal.h
@@ -85,39 +85,24 @@
 #ifdef WITH_OPENSSL
 # ifdef HAVE_EVP_SHA256
 #  define KEX_CURVE25519_METHODS \
-	"curve25519-sha256," \
-	"curve25519-sha256@libssh.org,"
+	"curve25519-sha256,"
 # else
 #  define KEX_CURVE25519_METHODS ""
 # endif
 #define KEX_COMMON_KEX \
 	KEX_CURVE25519_METHODS \
-	KEX_ECDH_METHODS \
-	KEX_SHA2_METHODS
+	KEX_ECDH_METHODS
 
 #define KEX_SERVER_KEX KEX_COMMON_KEX \
-	"diffie-hellman-group-exchange-sha1," \
-	KEX_SHA2_GROUP14 \
-	"diffie-hellman-group14-sha1," \
-	"diffie-hellman-group1-sha1"
+	"diffie-hellman-group14-sha1"
 
 #define KEX_CLIENT_KEX KEX_COMMON_KEX \
-	"diffie-hellman-group-exchange-sha1," \
-	KEX_SHA2_GROUP14 \
-	"diffie-hellman-group14-sha1," \
-	"diffie-hellman-group1-sha1"
+	"diffie-hellman-group14-sha1"
 
 #define	KEX_DEFAULT_PK_ALG	\
-	HOSTKEY_ECDSA_CERT_METHODS \
-	"ssh-ed25519-cert-v01@openssh.com," \
-	"ssh-rsa-cert-v01@openssh.com," \
-	"ssh-dss-cert-v01@openssh.com," \
 	HOSTKEY_ECDSA_METHODS \
 	"ssh-ed25519," \
-	"rsa-sha2-512," \
-	"rsa-sha2-256," \
-	"ssh-rsa," \
-	"ssh-dss"
+	"ssh-rsa"
 
 #define	KEX_FIPS_PK_ALG	\
 	HOSTKEY_ECDSA_CERT_METHODS \
@@ -130,22 +115,12 @@
 /* the actual algorithms */
 
 #define KEX_CLIENT_ENCRYPT \
-	"chacha20-poly1305@openssh.com," \
-	"aes128-ctr,aes192-ctr,aes256-ctr" \
-	AESGCM_CIPHER_MODES "," \
-	"aes128-cbc,aes192-cbc,aes256-cbc"
+	"aes128-ctr,aes256-ctr" \
+	AESGCM_CIPHER_MODES
 
-#define KEX_SERVER_ENCRYPT KEX_CLIENT_ENCRYPT "," \
-	"blowfish-cbc,cast128-cbc,3des-cbc"
+#define KEX_SERVER_ENCRYPT KEX_CLIENT_ENCRYPT
 
 #define KEX_SERVER_MAC \
-	"umac-64-etm@openssh.com," \
-	"umac-128-etm@openssh.com," \
-	"hmac-sha2-256-etm@openssh.com," \
-	"hmac-sha2-512-etm@openssh.com," \
-	"hmac-sha1-etm@openssh.com," \
-	"umac-64@openssh.com," \
-	"umac-128@openssh.com," \
 	"hmac-sha2-256," \
 	"hmac-sha2-512," \
 	"hmac-sha1"
